video: st_mipid02: addition of ST MIPID02 CSI bridge

avolmat-st opened the pull request:
The ST-MIPID02 is a CSI deserializer (up to 2 lanes) allowing to output up to 12 bit parallel interface.
It is present on the STM32MP135F discovery board and is part of the camera pipeline for this board:

GC2145 (sensor CSI) -> ST-MIPID02 -> DCMIPP (parallel interface)

This PR includes the driver and test entry for build_all. It will be added within the stm32mp135f-dk once the dcmipp PR #89407 will be merged.

This PR depends on the CCI interface PR #87935 as well as the GC2145-CSI PR #89571 in order to have the VIDEO_MIPI_ macros.

Commit:
video: st_mipid02: addition of ST MIPID02 CSI bridge

Addition of the support for the CSI to DVP bridge ST MIPID02.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: addition of CSI2 standard data-type defines

Add standard data-type macros within the video.h in order
to avoid having to redefine them all in each and every CSI
based driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented on drivers/video/Kconfig.st_mipid02:
```diff
@@ -0,0 +1,24 @@
+
+if VIDEO_ST_MIPID02
+
+config VIDEO_ST_MIPID02_CAPS_HEAP_SIZE
+	int "Size of a heap reserved to handle caps"
+	default 512
+	help
+	  Heap size reserved in order to handle caps. This could be
+	  reduced or increased depending on the amount of formats
+	  supported by the ST_MIPID02 source device.
```
Ultimately we would need a different API to avoid such issue, but it sounds like the right way to do in the meantime.

josuah commented on tests/drivers/build_all/video/app.overlay:
```diff
@@ -77,6 +84,37 @@
+						reg = <2>;
+
+						mipid02_2: endpoint {
+							bus-width = <8>;
+							hsync-active = <0>;
+							vsync-active = <0>;
+							pclk-sample = <0>;
```
It seems like these are not used anywhere but the hardware supports customizing them so makes sense to add them now.

Should there be `bus-type` too? Maybe not as it is not configurable.

avolmat-st replied:
I indeed forgot to add the support for controlling the syncho signals on the parallel side. Will first add them.

avolmat-st replied:
Synchro signal handling added. I actually removed the bus-width property from the app.overlay since it is not useful for the ST-MIPID02.

josuah reviewed (dismissed):
I could not spot blocking issues with it.
Interesting chip for Zephyr which has plenty of DVP-only boards...
Thank you for this driver!

avolmat-st commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,458 @@
+		LOG_ERR("No valid format desc available, should get/set_fmt prior to set_stream");
+		return -EIO;
+	}
+	link_freq = ctrl.val64 * video_bits_per_pixel(desc->csi_pixelformat) /
+		    (2 * cfg->csi.nb_lanes);
+
+	/* Enable lanes */
+	ret = video_write_cci_reg(&cfg->i2c, MIPID02_CLK_LANE_REG1,
+				  (2000000000 / link_freq) << MIPID02_CLK_LANE_REG1_UI_SHIFT |
+				  MIPID02_CLK_LANE_REG1_EN);
+	if (ret < 0) {
+		return ret;
+	}
```
No it doesn't here. Actually often the link_freq of the sensor is known, so in this case it is better to directly get the LINK_FREQ value from the sensor driver. When it isn't known it is calculated from the pixel_rate. This involve that the pixel_rate returned by the sensor is roughly right in order to get the correct link_freq.
Probably I could add first the LINK_FREQ control on the sensor side so that the if this one is available then we directly use it. There could be a helper for that to first try to pick the link_freq and in fallback if not available we use the pixel_rate to approximate it. Similar thing is done on Linux in fact. For the time being we don't have yet much CSI receiver in Zephyr so if we want to add stuff like that this might be a good time.

josuah replied:
Thank you, I did not know how it was all stitched together..

> if we want to add stuff like that this might be a good time

Putting it in a separate PR would make this one faster to merge, but also one more PR to wait for. Both work!

avolmat-st replied:
Hum, there are already several PRs which tend to get inter-dependent so I'll properly keep it like that for the moment to let those PRs get merged and will make a dedicated PR for such video-common general stuff, update required driver as well at same time.

ngphibang replied:
`Probably I could add first the LINK_FREQ control on the sensor side so that the if this one is available then we directly use it. There could be a helper for that to first try to pick the link_freq and in fallback if not available we use the pixel_rate to approximate it. Similar thing is done on Linux in fact. For the time being we don't have yet much CSI receiver in Zephyr so if we want to add stuff like that this might be a good time.`

I think this is a good occasion to add the link frequency control which is better than pass by the pixel rate. I didn't add it when adding pixel rate as there is no user of it at that time.

> Hum, there are already several PRs which tend to get inter-dependent so I'll properly keep it like that for the moment to let those PRs get merged and will make a dedicated PR for such video-common general stuff, update required driver as well at same time.

I think we can put it here in the same PR (as it is related stuff) or in a separate PR (to get merged faster), it doesn't matter. Inter-dependency is common and not a problem if we add the dependency tree in the PR description.

avolmat-st commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,458 @@
+		}
+	}
+
+	desc = mipid02_get_format_desc(drv_data->fmt.pixelformat);
+	if (!desc) {
+		return -EINVAL;
+	}
+
+	/* Get pixelrate information from source */
+	ret = video_get_ctrl(cfg->source_dev, &ctrl);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!desc) {
+		LOG_ERR("No valid format desc available, should get/set_fmt prior to set_stream");
+		return -EIO;
+	}
```
No ;) I modified this bit of code several times and missed that ;) Will correct.

Commit:
video: add LINK_FREQ ctrl and a helper to retrieve it

Add a ctrl VIDEO_CID_LINK_FREQ to indicate to a sink
the frequency at which a device streams data over
CSI2. Since not all source device currently provide the
LINK_FREQ control, add a helper function to retrieve
it, with a fall-back by doing an approximate via the
PIXEL_RATE control.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
I'm adding a new function video_get_link_frequency within the video-common in order to get the link-frequency from the source. If the link-frequency ctrl is not available (which is the case for the time being), it falls back on the pixel-rate.

avolmat-st commented:
New version pushed with PR #87935 commit added first in order to allow proper build of the PR.

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -78,6 +78,7 @@ static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32
 		*type = VIDEO_CTRL_TYPE_MENU;
 		break;
 	case VIDEO_CID_PIXEL_RATE:
+	case VIDEO_CID_LINK_FREQUENCY:
 		*type = VIDEO_CTRL_TYPE_INTEGER64;
 		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
 		break;
```
I think that link frequency control is rather an array of values (integer menu), no ? This is one of the main differences and adavantages between it and the pixel rate. And we can specify specify the supported link frequencies in the DT like:

```
&camera {
    port {
        camera_ep: endpoint {
            clock-lanes = <0>;
            data-lanes = <1 2>;
            link-frequencies = /bits/ 64 <400000000 800000000>; // In Hz
        };
    };
```
};

It's because the link frequency is board-specific characteristics and for some sensors, the master clock is provided by the upstream device.

This way, sensor driver will get the link frequency from the DT property and create the menu.

But if it seems complicated and not needed for the current usecase, we can do it later.

ngphibang replied:
I think it's fine like that, as long as the ctrl type is consistent, other things can be added and improved later. Thanks

Commit:
video: common: addition of INTEGER_MENU control type

Add a new INTEGER_MENU type allowing to store signed 64bits
integer into a menu.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
New version, with LINK_FREQUENCY now being a MENU_INTEGER type. Actually this type was not available yet, only MENU (strings) were available hence I've added a new intermediate commit to add this one prior to actually adding the new control itself. Now the helper is relying on this new type, queryctrl etc to figure out the link frequency value.

I've also update my gc2145 driver to work like that. If this one (ST-MIPID02) looks right I will push the GC2145 on top of this branch then.

avolmat-st commented:
I saw in the code that for the menu type it says that this is for standard or driver defined control. However I have a doubt about that since in case of ctrl which are not standard, the set_type_flag function is going to fallback to INTEGER type ctrl, hence not being possible to define custom menu ctrl I think. In order to achieve this it would actually be necessary to add a new parameter to the init functions if it is not a known ctrl.

This is not really related to this PR itself in fact.

ngphibang commented on drivers/video/video_ctrls.h:
```diff
@@ -64,6 +69,9 @@ int video_init_ctrl(struct video_ctrl *ctrl, const struct device *dev, uint32_t
 int video_init_menu_ctrl(struct video_ctrl *ctrl, const struct device *dev, uint32_t id,
 			 uint8_t def, const char *const menu[]);
 
+int video_init_integer_menu_ctrl(struct video_ctrl *ctrl, const struct device *dev, uint32_t id,
```
nit: `video_init_int_menu_ctrl` just to shorten the function name a bit

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -505,10 +535,15 @@ void video_print_ctrl(const struct device *const dev, const struct video_ctrl_qu
 		}
+	} else if (cq->type == VIDEO_CTRL_TYPE_MENU_INTEGER && cq->int_menu) {
+		while (cq->int_menu[i]) {
+			LOG_INF("%*s %u: %lld", 32, "", i, cq->int_menu[i]);
```
For string menu, the string name max length is 32. For integer menu, may be set to other value ? I don't know, it can be derived from a possible maximum link frequency (x Mhz) ?

avolmat-st replied:
For link frequency, I think the maximum would be around few GHz, but link-frequency is only 1 of the possible integer menu I think (while being probably a very big value considering that). Hence probably 12 or 16 should be suffisant in such case. What do you think ?

ngphibang replied:
Yes, link frequency seems the largest one. I think 12 is sufficient.

ngphibang reviewed (commented):
The integer menu controls LGTM. Thanks !

ngphibang commented:
> I saw in the code that for the menu type it says that this is for standard or driver defined control. However I have a doubt about that since in case of ctrl which are not standard, the set_type_flag function is going to fallback to INTEGER type ctrl, hence not being possible to define custom menu ctrl I think. In order to achieve this it would actually be necessary to add a new parameter to the init functions if it is not a known ctrl.

"Driver-defined menu" means the driver define the menu items but the control type is known, it does not mean a private & driver-defined menu control (I didn't see such a control, in Linux). Two examples are the test_pattern (driver-defined item) and the power line frequency (standard):

```
	ret = video_init_menu_ctrl(&ctrls->light_freq, dev, VIDEO_CID_POWER_LINE_FREQUENCY,
				   VIDEO_CID_POWER_LINE_FREQUENCY_50HZ, NULL);

	ret = video_init_menu_ctrl(&ctrls->test_pattern, dev, VIDEO_CID_TEST_PATTERN, 0,
				   test_pattern_menu);
```

avolmat-st commented:
> > I saw in the code that for the menu type it says that this is for standard or driver defined control. However I have a doubt about that since in case of ctrl which are not standard, the set_type_flag function is going to fallback to INTEGER type ctrl, hence not being possible to define custom menu ctrl I think. In order to achieve this it would actually be necessary to add a new parameter to the init functions if it is not a known ctrl.
> 
> "Driver-defined menu" means the driver defines the menu items but the control id is known, it does not mean a private & driver-defined menu control (I didn't see such a control, in Linux). Two examples are the test_pattern (driver-defined item) and the power line frequency (standard):
> 
> ```
> 	ret = video_init_menu_ctrl(&ctrls->light_freq, dev, VIDEO_CID_POWER_LINE_FREQUENCY,
> 				   VIDEO_CID_POWER_LINE_FREQUENCY_50HZ, NULL);
> 
> 	ret = video_init_menu_ctrl(&ctrls->test_pattern, dev, VIDEO_CID_TEST_PATTERN, 0,
> 				   test_pattern_menu);
> ```

Understood. Right I kind of read it as driver defined control ;) OK then. Thanks for the clarification.

ngphibang reviewed (commented):
LGTM, just wait for the dependency (CCI PR) get merged first

avolmat-st commented:
@ngphibang I am having kind of second thought about the link frequency helper. Should we really have that or should we simply enforce that any CSI driver MUST expose a LINK_FREQUENCY control. Currently, with the addition of the GC2145 csi, I think that is only 2 sensors being able to work in CSI. I will add the LINK_FREQUENCY ctrl to the GC2145 so only the OV5640 won't have. I guess we could add it. If we decide to enforce the requirement of LINK_FREQUENCY ctrl that helper doesn't really make sense. What do you think ?

ngphibang commented:
> @ngphibang I am having kind of second thought about the link frequency helper. Should we really have that or should we simply enforce that any CSI driver MUST expose a LINK_FREQUENCY control. Currently, with the addition of the GC2145 csi, I think that is only 2 sensors being able to work in CSI. I will add the LINK_FREQUENCY ctrl to the GC2145 so only the OV5640 won't have. I guess we could add it. If we decide to enforce the requirement of LINK_FREQUENCY ctrl that helper doesn't really make sense. What do you think ?

I think both are relevant and describe different aspects of things. Pixel rate is simpler to be used. Personally, I would not enforce anything if it's technically not wrong (e.g. if someone finds that link frequency is better they can make change btw).

avolmat-st commented:
> > @ngphibang I am having kind of second thought about the link frequency helper. Should we really have that or should we simply enforce that any CSI driver MUST expose a LINK_FREQUENCY control. Currently, with the addition of the GC2145 csi, I think that is only 2 sensors being able to work in CSI. I will add the LINK_FREQUENCY ctrl to the GC2145 so only the OV5640 won't have. I guess we could add it. If we decide to enforce the requirement of LINK_FREQUENCY ctrl that helper doesn't really make sense. What do you think ?
> 
> I think both are relevant and describe different aspects of things. Pixel rate is simpler to be used. Personally, I would not enforce anything if it's technically not wrong (e.g. if someone finds that link frequency is better they can make change btw).

To me the real important value in case of CSI is the LINK FREQUENCY since this is that link frequency which is related to the UI necessary for configuration of a CSI PHY. From the pixel rate we can only kind of guess a link frequency "enveloppe" that would be necessary to transfer THAT amount of pixel but that's not necessarily true and the exact value. How it is transmitted and the amount of pixels are 2 different aspects I think.
However as you said, if this turns out not to be used or turns out it should be there later on we can remove it ;) So let's not spend too much time on this now.

@loicpoulain the MIPI CSI2 DT header is not part of this serie instead of the GC2145 serie (since it is also necessary for the ST-MIPI which is a dependency for the GC2145 since it introduce the LINK_FREQUENCY ctrl).

josuah commented:
I opened this issue to help keeping track of everywhere link frequency (and other related API changes) are discussed:
- https://github.com/zephyrproject-rtos/zephyr/issues/90099

josuah commented on drivers/video/video_ctrls.h:
```diff
@@ -27,10 +27,12 @@ enum video_ctrl_type {
 	VIDEO_CTRL_TYPE_MENU = 4,
 	/** String type */
 	VIDEO_CTRL_TYPE_STRING = 5,
+	/** Menu integer type, standard or driver-defined menu */
+	VIDEO_CTRL_TYPE_MENU_INTEGER = 6,
```
Non-blocking as Linux controls are also organized slightly differently.

avolmat-st replied:
Moved to INTEGER_MENU

josuah commented on drivers/video/video_ctrls.h:
```diff
@@ -54,7 +56,10 @@ struct video_ctrl {
-	const char *const *menu;
+	union {
+		const char *const *menu;
+		const int64_t *int_menu;
```
Is it making more sense as `int64_menu`? Maybe not as it's to be used for both `int32_t` and `int64_t`.
If some day there is a kind of long int_menu (i.e. calibration data), then having `int_menu` free for it can help.

josuah commented on drivers/video/video_common.c:
```diff
@@ -342,3 +343,43 @@ int video_write_cci_multiregs16(const struct i2c_dt_spec *i2c, const struct vide
+	struct video_ctrl_query ctrl_query = {
+		.id = VIDEO_CID_LINK_FREQUENCY,
+	};
+
+
```
probably a typo: two empty lines

josuah commented on drivers/video/video_common.c:
```diff
@@ -342,3 +343,43 @@ int video_write_cci_multiregs16(const struct i2c_dt_spec *i2c, const struct vide
+fallback:
+	/* If VIDEO_CID_LINK_FREQUENCY is not available, approximate from VIDEO_CID_PIXEL_RATE */
+	ctrl.id = VIDEO_CID_PIXEL_RATE;
+	ret = video_get_ctrl(dev, &ctrl);
```
Since we query the *current* pixel rate, should we go ahead and query the current FOURCC and derive the `bits_per_pixel` from it? Just for the sake of ease of use.

avolmat-st replied:
For the time being this is only used by a single driver so maybe wait to have it used more widely and then see if, based on the usage of other driver we can put more stuff in it.

josuah commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -171,6 +171,9 @@ enum video_exposure_auto_type {
 /** Selection of the type of test pattern to represent */
 #define VIDEO_CID_TEST_PATTERN (VIDEO_CID_IMAGE_PROC_CLASS_BASE + 3)
 
+/** Link frequency, applicable for the CSI2 based devices. This control is read-only. */
+#define VIDEO_CID_LINK_FREQUENCY (VIDEO_CID_IMAGE_PROC_CLASS_BASE + 4)
```
In Linux, `V4L2_CID_IMAGE_PROC_CLASS_BASE + 4` is [`V4L2_CID_DEINTERLACING_MODE`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/v4l2-controls.h?h=v6.15-rc6#n1233).

There is this comment above:
https://github.com/zephyrproject-rtos/zephyr/blob/b5934e8dfe5c8dd4c1a9f2221341aa46e3dce72a/include/zephyr/drivers/video-controls.h#L22-L28

So we can go low-effort mode and import the content unchanged (for the doc also?) :)

```suggestion
/** The frequency of the data bus (e.g. parallel or CSI-2). This control is read-only. */
#define VIDEO_CID_LINK_FREQ (VIDEO_CID_IMAGE_PROC_CLASS_BASE + 4)
```

avolmat-st replied:
From a licence point of view, is that alright to have stuff from Linux into Zephyr ? I understand that using same numbering or name could help when creating stuff on top of the CTRL API but is that ok ?
Regarding importing the doc as well, I don't think that is acceptable considering that this would really be copy of contents from an incompatible licence.

ngphibang replied:
I think the license on the specification is applied for the spec itself, i.e. we need to follow it when making another spec based on it (derivative works). When it comes to implementing a specification in code, I think it is not considered a derivative work of the specification, so the license is not applied. I think this is the purpose of a specification, i.e. everyone needs to follow it to make the implementation (e.g. each vendor implements the MPEG spec differently) ... how we can follow a spec without using it (?).

However, I think we need other experts to explain these things and advise how to do now, especially with the documentation.

josuah replied:
To split the PR and longer-term discussion, would it be a suitable compromise to keep the Linux CIDs but leave the documentation empty until this is resolved?

ngphibang replied:
Nice ! For the copyright, we need to keep the original copyright ?

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -819,6 +819,23 @@ void video_closest_frmival_stepwise(const struct video_frmival_stepwise *stepwis
+/**
+ * @brief Return the link-frequency advertised by a device
+ *
+ * Device exposing a CSI link will should advertise at least one of the following two controls:
+ *   VIDEO_CID_LINK_FREQUENCY
+ *   VIDEO_CID_PIXEL_RATE
+ *
+ * At first the helper will try read the VIDEO_CID_LINK_FREQUENCY and if not available will
+ * approximate the link-frequency from the VIDEO_CID_PIXEL_RATE value, taking into consideration
+ * the bits per pixel of the format and the number of lanes.
```
Some help with doc proofreading:
```suggestion
 * Device exposing a CSI link should advertise at least one of the following two controls:
 * - @ref VIDEO_CID_LINK_FREQUENCY
 * - @ref VIDEO_CID_PIXEL_RATE
  *
 * At first the helper will try read the @ref VIDEO_CID_LINK_FREQUENCY and if not available will
 * approximate the link-frequency from the @ref VIDEO_CID_PIXEL_RATE value, taking into consideration
 * the bits per pixel of the format and the number of lanes.
```

josuah commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,462 @@
+	return 0;
+}
+
+static int mipid02_get_fmt(const struct device *dev, enum video_endpoint_id ep,
```
The removal of endpoint ID is now merged \:D

josuah reviewed (commented):
Maybe you meant to use the same CID numbers as in Linux (which can have impact in [UVC extensions](https://github.com/zephyrproject-rtos/zephyr/issues/86943) if that becomes a thing).
Everything else is very minor.

avolmat-st commented:
Pushed updates based on recent review points (still open points related to adjustments related to Linux).
Next I push the same commits rebased on top of head in order to allow proper build.

avolmat-st commented:
Twister issue is related to the GC2145 node (used to connect to the MIPID02 in test). This should get fixed by itself if the GC2145 CSI PR get merged first.

avolmat-st commented:
Actually this PR and the GC2145 PR (89571) both depend on each other.
The MIPID02 needs the GC2145 update in order to properly put the test case and the GC2145 PR needs the LINK_FREQ and MIPI_DT type to be defined.

Turns out I will have to move back the LINK_FREQ defintion and MIPI_DT type defintion back into the GC2145 PR in order to be able to first merge the GC2145/CSI PR.

Maybe I will first wait to have all discussion closed in this thread prior to moving the commits to the GC2145/CSI PR. Sorry for the mess ...

josuah commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,456 @@
+
+	/* Enable lanes */
+	ret = video_write_cci_reg(&cfg->i2c, MIPID02_CLK_LANE_REG1,
+				  (2000000000 / link_freq) << MIPID02_CLK_LANE_REG1_UI_SHIFT |
```
[non-blocking]
If you wanted to try [`FIELD_PREP`](https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#gaa03c8b31bf67a097dd9f8153a04ef86b):
```suggestion
				  FIELD_PREP(MIPID02_CLK_LANE_REG1_UI_MASK, 2000000000 / link_freq) |
```

josuah commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,456 @@
+		.i2c = I2C_DT_SPEC_INST_GET(n),							\
+		.source_dev = SOURCE_DEV(n),							\
+		.reset_gpio = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {0}),			\
+		.csi.nb_lanes = DT_PROP_LEN(DT_INST_ENDPOINT_BY_ID(n, 0, 0), data_lanes),	\
+		.csi.lanes[0] = DT_PROP_BY_IDX(DT_INST_ENDPOINT_BY_ID(n, 0, 0),		\
+					       data_lanes, 0),					\
+		.csi.lanes[1] = COND_CODE_1(DT_PROP_LEN(DT_INST_ENDPOINT_BY_ID(n, 0, 0),	\
+							data_lanes),				\
+					    (0),						\
+					    (DT_PROP_BY_IDX(DT_INST_ENDPOINT_BY_ID(n, 0, 0),	\
+							    data_lanes, 1))),			\
```
[discussion]
Some day I should add a `VIDEO_CSI_DT_SPEC_INST_GET(n)`...

josuah reviewed (changes requested):
One might need some attention (possibly a bug?).

Everything else is more about feedback long-term, and less important details IMHO.

Very interesting part! Hope this gets some public doc soon so we can use it outside ST's DKs. :)

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -365,3 +366,42 @@ int video_write_cci_multiregs16(const struct i2c_dt_spec *i2c, const struct vide
+
+int64_t video_get_link_frequency(const struct device *dev, uint8_t bpp, uint8_t lane_nb)
+{
+	int ret;
+
```
You meant extra blank line right ? If so, ok, I fixed it.

josuah replied:
There was a bug in the Github UI and I did not see the `int ret` either at that time. Thank you anyway!

avolmat-st commented on drivers/video/video_ctrls.c:
```diff
@@ -108,6 +108,10 @@ static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32
 		*type = VIDEO_CTRL_TYPE_INTEGER64;
 		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
 		break;
+	case VIDEO_CID_LINK_FREQ:
+		*type = VIDEO_CTRL_TYPE_MENU_INTEGER;
+		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+		break;
```
Yes when reading the control, it is equivalent to an integer type, it simply gives the index of the menu entry.

I am not sure about the requirement of having to switch link frequency at runtime. The approach is more that sensor would offer several possible link-frequencies and each platform will, within the device-tree mentions which link-frequency it can accept. Based on that the driver will adapt to configure itself in one or the other link-frequency.

Such mechanism is already available in Linux, we discussed a bit about that with @ngphibang previously but for now this isn't yet implemented in our helpers.

ngphibang replied:
> 1. application set a format
> 2. driver update the link-frequency menu that are available for this format
> 3. optionnal: application read the link-frequency, and if it has several link-frequency available, set the control to select the link-frequency it would prefer to use

Honestly, I didn't see myself in practice a model where the "driver updating the link frequency after application sets the format", it seems a bit complicated IMHO (?).

Some rare cases, I found simpler is the below:

https://elixir.bootlin.com/linux/v6.15/source/drivers/media/i2c/ov64a40.c#L36
https://docs.arducam.com/Raspberry-Pi-Camera/Native-camera/64MP-OV64A40/#resolution-frame-rates

Although in this case, it seems it can be done differently: changing the frame rate (with set_frmival) and set the link frequency accordingly but the driver chose to expose the changeable link frequencies.

So, it seems depending on the way the drivers want to model, it can be read-only or changeable but to be generic, v4l2 spec didn't force it read-only and let the drivers do it, I guess (?).

josuah replied:
I forgot about https://github.com/zephyrproject-rtos/zephyr/pull/89764#discussion_r2086152272 which already discussed about it. The ability to set read-only or not in each driver is same as Linux, so helps with inter-operability1

avolmat-st commented:
@josuah , updated based on your remarks. This is actually linked to the GC2145 / CSI PR, (tests can't be run without having GC2145 PR merged as well and GC2145 PR needs the LINK_FREQ CID which is in THIS PR), should I actually merge both PR here ? I could simply add the 3 GC2145 commits which enable CSI within this PR which would make the merge easier ?

The other approach is that I add only the relevant commits from this PR into of the GC2145 PR so that GC2145 PR can be merged first.

josuah commented:
> GC2145 PR needs the LINK_FREQ CID which is in THIS PR), should I actually merge both PR here ?
> I could simply add the 3 GC2145 commits which enable CSI within this PR which would make the merge easier ?

Putting the hardware-independent commits (CID, MIPI data types) might work. This should be quick as most of it is reviewed here already and these additions can also be useful in downstream applications.

> The other approach is that I add only the relevant commits from this PR into of the GC2145 PR so that GC2145 PR can be merged first.

If this is easy to move the commits around, then this works too! Whichever is most straightforward.

josuah commented on drivers/video/video_st_mipid02.c:
```diff
@@ -0,0 +1,456 @@
+	int ind = 0, caps_nb = 0;
+	int ret;
+
+	ret = video_get_caps(cfg->source_dev, caps);
```
Having a single driver type helps for simple devices that provide everything like Arducam Mega, or thermal sensor over SPI... But like you said it is "getting drivers more difficult" to implement. Reading Linux source will help me understand what is missing!

avolmat-st commented:
Updated test part, this PR can thus be merged without dependencies.

avolmat-st commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -362,6 +362,9 @@ enum video_camera_orientation {
  */
 #define VIDEO_CID_IMAGE_PROC_CLASS_BASE 0x009f0900
 
+/** Link frequency, applicable for the CSI2 based devices. This control is read-only. */
```
Indeed, I correct this and push it. Since those commits are also part of the GC2145/CSI PR (in order to be able to have CI OK, I also update on the GC2145/CSI PR as well).

ngphibang reviewed (approved):
LGTM, thanks !
